### ISSUE TITLE
[c++] Assess and improve code coverage

### ISSUE THREAD

**Issue by atolopko-czi** [single-cell-data/TileDB-SOMA#844]
[Split off from multi-language test coverage story: #778]

- [ ] Assess code coverage for C++ code base. If deemed acceptable, close this issue.
- [ ] If coverage can be improved, add necessary tests or enumerate in new issue(s) the areas where coverage can be reasonably improved. 

Helpful link: https://app.codecov.io/gh/single-cell-data/TileDB-SOMA/

Blocked on #1023.

**Comment by bekadavis9** [comment]
C++ code coverage sheet can be found here: https://docs.google.com/spreadsheets/d/1BqTv7v6MXRWcq4FQePYy9-BHVV6R6wDOf8nDEf-b4_I/edit?usp=sharing

Column A contains `SomaReader` APIs and Column B contains the Python invocations defined in `libtiledbsoma.cc`. The remaining columns represent each class in `libtiledbsoma/include/tiledbsoma`. Functions in a given row are called by one of the functions to their left. _Most_ functions are invoked within `SomaReader` or `ManagedQuery`, with some standalone tests. 
 
A green box denotes a standalone test, while orange indicates the function is not _exclusively_ tested. I believe that if a "far-left" function is green, anything to its right could be considered validated. I also found that a majority of the red functions are not entirely necessary and may be better off removed than validated.

**Comment by bekadavis9** [comment]
I also spent some time looking into improving the C++ Codecov reporting. It seems C++ implementations aren't as straightforward as with Python - generating the coverage report and sending it to Codecov requires some compile-time hacking. 

For CMake, Codecov themselves recommend using this: https://github.com/RWTH-HPC/CMake-codecov

**Comment by eddelbuettel** [comment]
We can get C++ coverage as a side effect of R coverage (modulo those parts for pybind11 R doesn't use).
